IntersectionObserver intersectionRatio < 1 observed

vitorroriz opened the pull request:
#### b5701774e49ff045c17f405df13ad7d6edfd1f41
<pre>
IntersectionObserver intersectionRatio &lt; 1 observed
<a href="https://bugs.webkit.org/show_bug.cgi?id=243573">https://bugs.webkit.org/show_bug.cgi?id=243573</a>
&lt;rdar://98541023&gt;

Reviewed by Simon Fraser.

* Source/WebCore/dom/Document.cpp:
(WebCore::computeClippedRectInRootContentsSpace):
We call `computeClippedRectInRootContentsSpace` recursively until the renderer’s frame
 is the main frame. If we are not yet in the main frame, we use `contentsToView` to adjust
the target rect to the origin’s view scrolling position. Before this patch we were snapping
the target rect before passing it.

The patch removes the snapping, forcing the use of FloatRect overload of contentsToView
to avoid inaccurate calculation of intersection ratio when the target is not directly in the
main frame.

* LayoutTests/imported/w3c/web-platform-tests/intersection-observer/intersection-ratio-with-fractional-bounds-in-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/intersection-observer/intersection-ratio-with-fractional-bounds-in-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/intersection-observer/resources/intersection-ratio-with-fractional-bounds-in-iframe-content.html: Added.

Canonical link: <a href="https://commits.webkit.org/254244@main">https://commits.webkit.org/254244@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/27cc2269a8371065999ee866fb479b39a2d70577

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88385 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32824 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19221 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97567 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/153040 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92349 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/31325 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26972 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80579 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92225 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93992 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24914 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/75234 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24889 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79814 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79934 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67853 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/28937 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13914 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/28928 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14937 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2985 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/32118 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37853 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31067 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34050 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/imported/w3c/web-platform-tests/intersection-observer/intersection-ratio-with-fractional-bounds-in-iframe.html:
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <script src="../resources/testharness.js"></script>
```
These should be:
```
<script src="/resources/testharness.js"></script>
<script src="/resources/testharnessreport.js"></script>
```

vitorroriz committed:
IntersectionObserver intersectionRatio < 1 observed
https://bugs.webkit.org/show_bug.cgi?id=243573
<rdar://98541023>

Reviewed by Simon Fraser.

* Source/WebCore/dom/Document.cpp:
(WebCore::computeClippedRectInRootContentsSpace):
We call `computeClippedRectInRootContentsSpace` recursively until the renderer’s frame
 is the main frame. If we are not yet in the main frame, we use `contentsToView` to adjust
the target rect to the origin’s view scrolling position. Before this patch we were snapping
the target rect before passing it.

The patch removes the snapping, forcing the use of FloatRect overload of contentsToView
to avoid inaccurate calculation of intersection ratio when the target is not directly in the
main frame.

* LayoutTests/imported/w3c/web-platform-tests/intersection-observer/intersection-ratio-with-fractional-bounds-in-iframe-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/intersection-observer/intersection-ratio-with-fractional-bounds-in-iframe.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/intersection-observer/resources/intersection-ratio-with-fractional-bounds-in-iframe-content.html: Added.

Canonical link: https://commits.webkit.org/254244@main
